src/Workspaces/Core/Portable/Utilities/IReadOnlyListExtensions.cs

```diff
@@ -1,12 +1,38 @@
-﻿using System.Collections.Generic;
+﻿using System.Collections;
+using System.Collections.Generic;
 
 namespace Microsoft.CodeAnalysis.Utilities
 {
     internal static class IReadOnlyListExtensions
     {
+        public static IReadOnlyList<T> ToReadOnlyList<T>(this IList<T> list)
+        {
+            if (list is IReadOnlyList<T> readOnlyList)
+            {
+                return readOnlyList;
+            }
+
+            return new ReadOnlyList<T>(list);
+        }
+
         public static T Last<T>(this IReadOnlyList<T> list)
         {
             return list[list.Count - 1];
         }
+
+        private class ReadOnlyList<T> : IReadOnlyList<T>
```

Comment:
IReadOnlyList has indexer which is drop in replacement for ImmutebleArray, IreadOnlyCollection doesn't have indexer.